File: projectBlueWater/src/main/java/com/lasthopesoftware/bluewater/client/playback/engine/selection/PlaybackEngineTypeSelectionPersistence.java
package com.lasthopesoftware.bluewater.client.playback.engine.selection;

import android.content.Context;
import android.preference.PreferenceManager;

import com.lasthopesoftware.bluewater.ApplicationConstants;
import com.lasthopesoftware.bluewater.client.playback.engine.selection.broadcast.PlaybackEngineTypeChangedBroadcaster;

public class PlaybackEngineTypeSelectionPersistence implements SelectPlaybackEngineType {

	private final Context context;
	private final PlaybackEngineTypeChangedBroadcaster playbackEngineTypeChangedBroadcaster;

	public PlaybackEngineTypeSelectionPersistence(Context context, PlaybackEngineTypeChangedBroadcaster playbackEngineTypeChangedBroadcaster) {
		this.context = context;
		this.playbackEngineTypeChangedBroadcaster = playbackEngineTypeChangedBroadcaster;
	}

	@Override
	public void selectPlaybackEngine(PlaybackEngineType playbackEngineType) {
		PreferenceManager
			.getDefaultSharedPreferences(context).edit()
			.putString(ApplicationConstants.PreferenceConstants.playbackEngine, playbackEngineType.name())
			.apply();

		playbackEngineTypeChangedBroadcaster.broadcastPlaybackEngineTypeChanged(playbackEngineType);
	}
}

File: projectBlueWater/src/main/java/com/lasthopesoftware/bluewater/client/playback/engine/selection/PlaybackEngineTypeSelectionPersistence.kt
package com.lasthopesoftware.bluewater.client.playback.engine.selection

import com.lasthopesoftware.bluewater.client.playback.engine.selection.broadcast.PlaybackEngineTypeChangedBroadcaster
import com.lasthopesoftware.bluewater.features.access.HoldApplicationFeatureConfiguration
import com.lasthopesoftware.promises.extensions.unitResponse
import com.namehillsoftware.handoff.promises.Promise

class PlaybackEngineTypeSelectionPersistence(private val applicationFeatureConfiguration: HoldApplicationFeatureConfiguration, private val playbackEngineTypeChangedBroadcaster: PlaybackEngineTypeChangedBroadcaster) : SelectPlaybackEngineType {
	override fun selectPlaybackEngine(playbackEngineType: PlaybackEngineType): Promise<Unit> =
		applicationFeatureConfiguration.promiseFeatureConfiguration()
			.eventually { s ->
				if (s.playbackEngineType != playbackEngineType) {
					playbackEngineTypeChangedBroadcaster.broadcastPlaybackEngineTypeChanged(playbackEngineType)
				}

				val newConfiguration = s.copy(playbackEngineType = playbackEngineType)
				applicationFeatureConfiguration.promiseUpdatedFeatureConfiguration(newConfiguration)
			}
			.unitResponse()
}
